#### Дополнительные задачи для самостоятельной практики к теме 1.2

#### Задание 1. Может ли число всех событий вероятностного пространства быть равным 65; 66; 64?

Рассмотрим, что означает "число всех событий". 

Если $\Omega$ — конечное пространство элементарных исходов и $|\Omega| = n$, 
то общее число событий (подмножеств множества $\Omega$) равно:

$$
|2^\Omega| = 2^n
$$

То есть число **всех событий** всегда степень двойки.

Проверим, могут ли 64, 65, 66 быть такими значениями.


In [4]:
# Проверка, являются ли данные числа степенями двойки
def is_power_of_two(x):
    return x > 0 and (x & (x - 1)) == 0

numbers = [64, 65, 66]
results = {n: is_power_of_two(n) for n in numbers}
results


{64: True, 65: False, 66: False}

**Вывод:**
- Только число 64 — допустимое число всех событий.
- 65 и 66 не могут быть числом всех событий, так как не являются степенями двойки.


#### Задание 2. Выборка из урны (m шаров, n извлечений)

Рассматриваются два способа:
- **С возвращением**
- **Без возвращения**

#### A) Упорядоченная выборка

##### 1) С возвращением:
Каждое извлечение — выбор из m элементов, всего n раз:
$$
|\Omega| = m^n
$$

##### 2) Без возвращения:
На каждом шаге количество доступных элементов уменьшается:
$$
|\Omega| = \frac{m!}{(m - n)!}
$$

#### B) Неупорядоченная выборка

##### 1) С возвращением:
Элементы можно выбирать повторно, порядок не важен.
Это сочетание с повторениями:
$$
|\Omega| = C_{m + n - 1}^n = \binom{m + n - 1}{n}
$$

##### 2) Без возвращения:
Классическое сочетание:
$$
|\Omega| = C_m^n = \binom{m}{n}
$$


In [10]:
import math

def combinations(n, k):
    return math.comb(n, k)

def permutations(m, n):
    return math.perm(m, n)

# Пример: m = 10, n = 3
m, n = 10, 3

ordered_with_replacement = m ** n
ordered_without_replacement = permutations(m, n)

unordered_with_replacement = combinations(m + n - 1, n)
unordered_without_replacement = combinations(m, n)

{
    "Упорядоченная, с возвращением": ordered_with_replacement,
    "Упорядоченная, без возвращения": ordered_without_replacement,
    "Неупорядоченная, с возвращением": unordered_with_replacement,
    "Неупорядоченная, без возвращения": unordered_without_replacement
}


{'Упорядоченная, с возвращением': 1000,
 'Упорядоченная, без возвращения': 720,
 'Неупорядоченная, с возвращением': 220,
 'Неупорядоченная, без возвращения': 120}

#### Задание 3. Подбрасывание монеты до r-го герба

Честная монета подбрасывается, пока не выпадет $r$ гербов.

Каждое подбрасывание — результат: "H" (герб) или "T" (решка).

Событие: **эксперимент заканчивается после n-го подбрасывания** означает:
- в первых $(n - 1)$ подбрасываниях — ровно $(r - 1)$ гербов,
- на $n$-м подбрасывании — обязательно герб.

Нужно:
- посчитать, сколько строк длины $n$, в которых $r - 1$ гербов в первых $n - 1$ позициях
- последний — герб

Это эквивалентно:
$$
\text{Число таких последовательностей} = \binom{n - 1}{r - 1}
$$

Подставим $n = 20$, $r = 5$:


In [13]:
n = 20
r = 5

elementary_events = combinations(n - 1, r - 1)
elementary_events


3876

**Ответ:**
Количество элементарных исходов, при которых эксперимент заканчивается на 20-м подбрасывании, равно $$
\binom{19}{4} = 3876
$$
